Panic on connect when a client sends a non UTF-8 server address

Running 1.45.0 in kubernetes with the metrics backend set to prometheus.

Every so often the process panics and restarts when something connects with a garbage server address in the handshake. Fairly sure it is just internet scanners hitting the exposed port, the addresses are never anything sensible.

Worth saying up front that the client IP in the log below is not the real source. My load balancer does a plain TCP passthrough to mc-router, so every connection shows up as the load balancer pod IP and I cannot currently tell who sent it. That also means the load balancer is only the conduit here, it hands the bytes over untouched rather than mangling them.

```
time="2026-07-26T16:56:01Z" level=info msg="Connecting to backend" backendHostPort="mc-router-minecraft-fallback:25565" client="10.112.0.244:46562" player= server="\x01\xd9 4\xc3X\xdc\v\x85\xfek\xfd;s\a\x11\xb49KC-"
time="2026-07-26T16:56:01Z" level=info msg="Closed connection to backend" backendHostPort="mc-router-minecraft-fallback:25565" client="10.112.0.244:46562" connectionCount=1 player=
panic: label server_address: value "\x01\xd9 4\xc3X\xdc\v\x85\xfek\xfd;s\a\x11\xb49KC-" is not valid UTF-8

goroutine 130471 [running]:
github.com/prometheus/client_golang/prometheus.(*GaugeVec).With(0x13b6b40?, 0x3f24e3bbbf0?)
        /home/runner/go/pkg/mod/github.com/prometheus/client_golang@v1.24.0/prometheus/gauge.go:252 +0x51
github.com/go-kit/kit/metrics/prometheus.(*Gauge).Set(0x3f24df405c0, 0x3ff0000000000000)
        /home/runner/go/pkg/mod/github.com/go-kit/kit@v0.13.0/metrics/prometheus/prometheus.go:79 +0xff
github.com/itzg/mc-router/server.(*Connector).findAndConnectBackend(0x3f24da985a0, {0x165bc38, 0x3f24df82050}, {0x1649170, 0x3f24e00a3c0}, {0x1643460, 0x3f24df6c150}, {0x3f24dba61a0, 0x15}, 0x0, ...)
        /home/runner/work/mc-router/mc-router/server/connector.go:726 +0x2763
github.com/itzg/mc-router/server.(*Connector).HandleConnection(0x3f24da985a0, {0x165bc38, 0x3f24df82050})
        /home/runner/work/mc-router/mc-router/server/connector.go:365 +0x1a68
created by github.com/itzg/mc-router/server.(*Connector).acceptConnections in goroutine 37
        /home/runner/work/mc-router/mc-router/server/connector.go:256 +0x26f
```

In case it helps, that address is 21 bytes of binary rather than a mangled hostname:

```
01 d9 20 34 c3 58 dc 0b 85 fe 6b fd 3b 73 07 11 b4 39 4b 43 2d
```

Roughly a third of the bytes have the high bit set and it does not match the start of TLS, HTTP, SSH or PROXY protocol, so it looks like random or encrypted junk from something scanning the port.

As far as I can tell (and some help from Claude) the server address from the handshake goes straight into a prometheus label here:

https://github.com/itzg/mc-router/blob/master/server/connector.go#L540-L542

client_golang validates that label values are valid UTF-8 and `With()` panics when they are not, so anything that connects and sends non UTF-8 bytes can take the whole process down. Worth saying this is not something new in client_golang 1.24, the same `utf8.ValidString` check is in 1.23 and older too, so I do not think the recent dependency bump caused it.

The annoying part is it kills every connected player, not just the connection that sent the bad data.

There is the same pattern a few lines further down for `player_name` and `player_uuid`, though I have not seen that one actually trigger.

For now I worked around it by setting the metrics backend to `discard`, which skips the label entirely, but that obviously loses all the metrics.

Would it make sense to validate or sanitize the label value before passing it in, or just skip recording that metric when the address is not valid UTF-8? Happy to test a patch if that helps.

--- a/mcproto/read.go
+++ b/mcproto/read.go
@@ -13,6 +13,8 @@ import (
 
 	"github.com/google/uuid"
 
+	"unicode/utf8"
+
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"golang.org/x/text/encoding/unicode"
@@ -183,7 +185,12 @@ func ReadUTF16BEString(reader io.Reader, symbolLen uint16) (string, error) {
 		return "", err
 	}
 
-	return string(result), nil
+	str := string(result)
+	if !utf8.ValidString(str) {
+		return "", errors.New("string is not valid UTF-8")
+	}
+
+	return str, nil
 }
 
 func ReadFrame(reader io.Reader, addr net.Addr) (*Frame, error) {
@@ -301,7 +308,12 @@ func ReadString(reader io.Reader) (string, error) {
 		strBuilder.WriteByte(b[0])
 	}
 
-	return strBuilder.String(), nil
+	str := strBuilder.String()
+	if !utf8.ValidString(str) {
+		return "", errors.New("string is not valid UTF-8")
+	}
+
+	return str, nil
 }
 
 func ReadByte(reader io.Reader) (byte, error) {

--- a/mcproto/read_test.go
+++ b/mcproto/read_test.go
@@ -16,6 +16,28 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func TestReadString(t *testing.T) {
+	t.Run("Valid UTF-8", func(t *testing.T) {
+		var buf bytes.Buffer
+		_ = WriteString(&buf, "mc.example.com")
+		str, err := ReadString(&buf)
+		require.NoError(t, err)
+		assert.Equal(t, "mc.example.com", str)
+	})
+
+	t.Run("Invalid UTF-8", func(t *testing.T) {
+		var buf bytes.Buffer
+		invalidBytes := []byte{0x01, 0xd9, 0x20, 0x34, 0xc3, 0x58, 0xdc, 0x0b, 0x85, 0xfe, 0x6b, 0xfd, 0x3b, 0x73, 0x07, 0x11, 0xb4, 0x39, 0x4b, 0x43, 0x2d}
+		_ = WriteVarInt(&buf, int32(len(invalidBytes)))
+		buf.Write(invalidBytes)
+
+		str, err := ReadString(&buf)
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "not valid UTF-8")
+		assert.Empty(t, str)
+	})
+}
+
 func TestReadVarInt(t *testing.T) {
 	tests := []struct {
 		Name     string

--- a/server/connector.go
+++ b/server/connector.go
@@ -8,9 +8,11 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
+	"unicode/utf8"
 
 	"golang.ngrok.com/ngrok"
 	"golang.ngrok.com/ngrok/config"
@@ -524,6 +526,13 @@ func (c *Connector) readPlayerInfo(protocolVersion mcproto.ProtocolVersion, buff
 	}
 }
 
+func sanitizeUTF8(s string) string {
+	if utf8.ValidString(s) {
+		return s
+	}
+	return strings.ToValidUTF8(s, "")
+}
+
 func (c *Connector) cleanupBackendConnection(clientAddr net.Addr, serverAddress string, playerInfo *PlayerInfo, backendHostPort string, scalingTarget string, cleanupMetrics bool, checkScaleDown bool) {
 	if c.connectionNotifier != nil {
 		err := c.connectionNotifier.NotifyDisconnected(c.ctx, clientAddr, serverAddress, playerInfo, backendHostPort)
@@ -538,16 +547,16 @@ func (c *Connector) cleanupBackendConnection(clientAddr net.Addr, serverAddress 
 
 		c.activeConnections.Decrement(backendHostPort)
 		c.metrics.ServerActiveConnections.
-			With("server_address", serverAddress).
+			With("server_address", sanitizeUTF8(serverAddress)).
 			Set(float64(c.activeConnections.GetCount(backendHostPort)))
 
 		c.scaleActiveConnections.Decrement(scalingTarget)
 
 		if c.recordLogins && playerInfo != nil {
 			c.metrics.ServerActivePlayer.
-				With("player_name", playerInfo.Name).
-				With("player_uuid", playerInfo.Uuid.String()).
-				With("server_address", serverAddress).
+				With("player_name", sanitizeUTF8(playerInfo.Name)).
+				With("player_uuid", sanitizeUTF8(playerInfo.Uuid.String())).
+				With("server_address", sanitizeUTF8(serverAddress)).
 				Set(0)
 		}
 	}
@@ -714,7 +723,7 @@ func (c *Connector) findAndConnectBackend(frontendConn net.Conn,
 		}
 	}
 
-	c.metrics.ConnectionsBackend.With("host", resolvedHost).Add(1)
+	c.metrics.ConnectionsBackend.With("host", sanitizeUTF8(resolvedHost)).Add(1)
 
 	c.metrics.ActiveConnections.Set(float64(
 		atomic.AddInt32(&c.totalActiveConnections, 1)))
@@ -722,7 +731,7 @@ func (c *Connector) findAndConnectBackend(frontendConn net.Conn,
 	c.activeConnections.Increment(backendHostPort)
 	c.scaleActiveConnections.Increment(scalingTarget)
 	c.metrics.ServerActiveConnections.
-		With("server_address", serverAddress).
+		With("server_address", sanitizeUTF8(serverAddress)).
 		Set(float64(c.activeConnections.GetCount(backendHostPort)))
 
 	if c.recordLogins && playerInfo != nil {
@@ -733,15 +742,15 @@ func (c *Connector) findAndConnectBackend(frontendConn net.Conn,
 			Info("Player attempted to login to server")
 
 		c.metrics.ServerActivePlayer.
-			With("player_name", playerInfo.Name).
-			With("player_uuid", playerInfo.Uuid.String()).
-			With("server_address", serverAddress).
+			With("player_name", sanitizeUTF8(playerInfo.Name)).
+			With("player_uuid", sanitizeUTF8(playerInfo.Uuid.String())).
+			With("server_address", sanitizeUTF8(serverAddress)).
 			Set(1)
 
 		c.metrics.ServerLogins.
-			With("player_name", playerInfo.Name).
-			With("player_uuid", playerInfo.Uuid.String()).
-			With("server_address", serverAddress).
+			With("player_name", sanitizeUTF8(playerInfo.Name)).
+			With("player_uuid", sanitizeUTF8(playerInfo.Uuid.String())).
+			With("server_address", sanitizeUTF8(serverAddress)).
 			Add(1)
 	}
 

--- a/server/connector_test.go
+++ b/server/connector_test.go
@@ -188,3 +188,37 @@ func TestConnectorMOTDFallback(t *testing.T) {
 	assert.Contains(t, jsonStr, "fallback asleep")
 	assert.False(t, scaleUpCalled, "Waker should NOT be called for a status request")
 }
+
+func TestConnectorNonUTF8Handshake(t *testing.T) {
+	routes := NewRoutes(t.Context())
+	downScaler := NewDownScaler(false, 5*time.Second)
+
+	metricsBuilder := prometheusMetricsBuilder{}
+	c := NewConnector(t.Context(), routes, downScaler, metricsBuilder.BuildConnectorMetrics(), false, false, nil)
+
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	defer ln.Close()
+
+	go c.acceptConnections(ln, 100, 0)
+
+	clientConn, err := net.Dial("tcp", ln.Addr().String())
+	require.NoError(t, err)
+	defer clientConn.Close()
+
+	invalidBytes := []byte{0x01, 0xd9, 0x20, 0x34, 0xc3, 0x58, 0xdc, 0x0b, 0x85, 0xfe, 0x6b, 0xfd, 0x3b, 0x73, 0x07, 0x11, 0xb4, 0x39, 0x4b, 0x43, 0x2d}
+	err = writeTestPacket(clientConn, 0x00, func(w io.Writer) {
+		_ = mcproto.WriteVarInt(w, 758)
+		_ = mcproto.WriteVarInt(w, int32(len(invalidBytes)))
+		w.Write(invalidBytes)
+		w.Write([]byte{0x63, 0xdd})
+		_ = mcproto.WriteVarInt(w, 1)
+	})
+	require.NoError(t, err)
+
+	// Wait briefly for connection processing to finish without panicking
+	_ = clientConn.SetReadDeadline(time.Now().Add(1 * time.Second))
+	buf := make([]byte, 10)
+	_, _ = clientConn.Read(buf)
+}
+

